### ISSUE TITLE
No parsing at all on some sites

### ISSUE THREAD

**Issue by Mickael-van-der-Beek** [fb55/htmlparser2#136]
On some URLs, ```htmlparser2``` doesn't parse the HTML at all.

To reproduce this bug, I created this little test case:

```
var request = require('request');
var htmlparser = require('htmlparser2');

var requestStream = request('https://news.osu.edu/');

var htmlParserStream = new (htmlparser.Parser)({
	onopentag: function (name, attrs) {
		console.log('[HTML-PARSER] onopentag=', arguments);
	},
	onclosetag: function (name, attrs) {
		console.log('[HTML-PARSER] onclosetag=', arguments);
	},
	onerror: function (e) {
		console.log('[HTML-PARSER] onerror=', arguments);
	},
	onend: function () {
		console.log('[HTML-PARSER] onend=', arguments);
	}
}, {
	lowerCaseAttributeNames: false,
	decodeEntities: true,
	lowerCaseTags: false
});

requestStream
	.on('data', function (data) {
		console.log('[REQUEST-STREAM] data=', arguments);
	})
	.on('end', function () {
		console.log('[REQUEST-STREAM] end=', arguments);
	})
	.on('close', function () {
		console.log('[REQUEST-STREAM] close=', arguments);
	})
	.pipe(htmlParserStream);
```

You will see that the following lines will get printed in the command line:

```
[REQUEST-STREAM] data= { '0': <Buffer 3c 21 44 4f 43 54 59 50 45 20 68 74 6d 6c 3e 0a 3c 21 2d 2d 5b 69 66 20 49 45 20 38 5d 3e 20 20 20 20 20 20 20 20 20 3c 68 74 6d 6c 20 63 6c 61 73 73 3d ...> }
[REQUEST-STREAM] data= { '0': <Buffer 74 65 22 3e 41 70 72 20 32 34 3c 2f 73 70 61 6e 3e 0a 09 20 20 20 09 3c 64 69 76 20 63 6c 61 73 73 3d 22 73 74 6f 72 79 5f 73 75 6d 6d 61 72 79 22 3e 0a ...> }
[REQUEST-STREAM] end= { '0': undefined }
[HTML-PARSER] onend= {}
```

So as you can see, data is correctly passed to ```htmlparser2``` who then simply calls the ```onend``` hook without any tags opened / closed or any errors.

I'm really not sure what could cause this problem since the website I tested this issue with, ```https://news.osu.edu/``` doesn't have that crazy of a markup.

**Comment by Mickael-van-der-Beek** [comment]
This seems to work for some reason:

```
var htmlparser = require('htmlparser2');
var request = require('request');
var fs = require('fs');

function getHtmlParser (index) {
	return new (htmlparser.Parser)({
		onopentag: function (name, attrs) {
			console.log('[HTML-PARSER' + index + '] onopentag=', arguments);
		},
		onclosetag: function (name, attrs) {
			console.log('[HTML-PARSER' + index + '] onclosetag=', arguments);
		},
		onerror: function (e) {
			console.log('[HTML-PARSER' + index + '] onerror=', arguments);
		},
		onend: function () {
			console.log('[HTML-PARSER' + index + '] onend=', arguments);
		}
	}, {
		lowerCaseAttributeNames: false,
		decodeEntities: true,
		lowerCaseTags: false
	});
}

var requestStream = request('https://news.osu.edu/');

var fileWriteStream = fs.createWriteStream('osu-page-dump.html', {
	flags: 'w',
	encoding: 'utf8'
});

requestStream
	.pipe(
		fileWriteStream
	);

requestStream
	.pipe(
		getHtmlParser(0)
	);

requestStream
	.on('error', function (error) {
		console.log('[REQUEST-STREAM] error=', arguments);
	})
	.on('data', function (data) {
		console.log('[REQUEST-STREAM] data=', arguments);
	})
	.on('end', function () {
		console.log('[REQUEST-STREAM] end=', arguments);
	})
	.on('close', function () {
		console.log('[REQUEST-STREAM] close=', arguments);
	});

var fileReadStream = fs.createReadStream('osu-page-dump.html', {
	flags: 'r',
	encoding: 'utf8'
});

fileWriteStream
	.on('finish', function () {
		console.log('[FILE-WRITE-STREAM] finish=', arguments);

		fileReadStream
			.pipe(
				getHtmlParser(1)
			);
	});
```

**Comment by Mickael-van-der-Beek** [comment]
I'm getting the same issue from this website: ```http://abcnews.go.com/```.

I might investigate this issue tonight if I have some free time.

**Comment by fb55** [comment]
It would be great if you could provide a simple test case, as webpages might change their content over time or by IP location. I'm hesitant to spend time trying to figure out what's wrong when I might not encounter the problem at all.

**Comment by Mickael-van-der-Beek** [comment]
Yup, np, I understand.

I did quite some testing and debugging to reduce my app to the minimum amount of lines of code and currently, although I can't provide you with a proof of concept yet, I can tell you that the issue or at least the difference comes from the ```request``` module.

In version ```request@2.53.0``` the script below will bug out and end without having parsed any HTML at all, but with ```request@2.60.0``` it will work fine.

```
var htmlparser2 = require('htmlparser2');
var request = require('request');

var requestHandlerStream = request({
	url: 'http://abcnews.go.com/',

	followRedirect: true,
	maxRedirects: 7,
	strictSSL: false,
	timeout: 15000,
	headers: {
		'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.115 Safari/537.36',

		'Accept-Encoding': 'gzip, deflate',
		'Accept-Language': 'en',
		'Accept': '*/*',

		'Cache-Control': 'no-cache',
		'Pragma': 'no-cache'
	},
	method: 'GET',

	pool: false,
	gzip: false
});

var contentHandlerStream = new (htmlparser2.Parser)({
	onopentag: function (name, attrs) {
		console.log('\n> [HTML_PARSER] NAME=', name, 'ATTRS=', attrs);
	},
	onerror: function (e) {
		console.log('\n> [HTML_PARSER] ERROR=', e);
	},
	onend: function () {
		console.log('\n> [HTML_PARSER] END');
	}
});

requestHandlerStream.on('response', function (res) {
	res.pipe(contentHandlerStream);
});

requestHandlerStream.on('error', function (e) {
	console.log('\n> [REQ] ERROR=', e);
});
```

**Comment by Mickael-van-der-Beek** [comment]
Why this site specifically bugged out rather than any other one remains a mystery and I'll have to investigate this later.

**Comment by Mickael-van-der-Beek** [comment]
Ok I binary-search through every version between ```request@2.53.0``` and ```request@2.60.0``` and the bug seems to disappear at version ```request@2.56.0```.

So somewhere in this diff:

https://github.com/request/request/compare/v2.55.1...v2.56.0

there's a fix to this situation.

At this stage though, it is still unclear to me if the bug comes from ```request``` or ```htmlparser2```.

**Comment by Mickael-van-der-Beek** [comment]
Ok I finally found the culprit. These are the lines that fixed the bug in ```request@2.56.0``` from ```request@2.55.1```:

https://github.com/request/request/commit/880d9a078665ee8d7dbda6cc26d58d7bebc2700d#diff-ccc0734f65dd7a299409ff07d35be095

and here is the linked issue / pull-request:

https://github.com/request/request/pull/1569

Now I can finally go to sleep ... ><'

**Comment by fb55** [comment]
Haha, nice find. Thanks for following up!
